Implement torch.i0

muthuArivoli opened the pull request:
Related to #38349

muthuArivoli committed:
implement bessel function

muthuArivoli committed:
implement kaiser

muthuArivoli committed:
add tests for i0

muthuArivoli committed:
bound tests

muthuArivoli committed:
add docs

muthuArivoli committed:
autograd

muthuArivoli committed:
attempt to fix cuda

muthuArivoli committed:
use c10 cuda compat

muthuArivoli committed:
try with function implementation in cuda

muthuArivoli committed:
fix casting

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-kaiser

muthuArivoli committed:
use doubles and test against scipy

muthuArivoli committed:
fix casting in cuda

muthuArivoli committed:
fix tests

muthuArivoli committed:
fix tests 2

muthuArivoli committed:
fix bfloat16

muthuArivoli committed:
fix float16 test

mruberry commented:
> @mruberry PTAL. This PR implements the kaiser window function. I have a few questions.

These are excellent questions.

> 1. The kaiser window function requires the modified Bessel function of the first kind, order 0 (i0). I went ahead and wrote an implementation for it, torch.i0, since numpy has the equivalent function np.i0. Is this alright, or should I remove the function?

That sounds correct but I'll take a closer look as I review (in a moment). 

> 2. For the i0 function, I followed the scipy.special.i0 implementation (a wrapper around the Cephes library function) as recommended in the np documentation. However, np supports complex numbers, while scipy.special does not, so should I support complex numbers? It shouldn't be too hard, I think numpy just takes the magnitude of the complex number.

Complex number support can be a separable follow-up. Let's focus on real values for the moment.

> 3. For testing, I modified the already existing test_signal_windows function. Should I create a an entirely new test and not touch the existing test, or is it alright that I modified the existing test?

Modifying an existing test is fine. I'll take a look at the details during the review.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -384,3 +384,120 @@ calc_gcd(T a, T b) {
 }
+
+/*
+ * The following function has been modified from code that comes with the following copyright notice.
```
There are multiple distributions of Cephes and its author, Stephen Moshier, has made clear that the code can be reused freely (and without requiring a particular license or copyright be distributed with it). I would add a comment document this function as normal, then add a sentence stating it's derived from an implementation in the Cephes Math Library.

mruberry replied:
I actually took some time to compare scipy.special.i0 with an approximation of the modified Bessel function of the first kind that's accurate in the limit, and I was impressed with how closely they matched on both smaller and larger values. In particular, on larger values I had to go out to 40 iterations of the approximating series before approach the value produced by scipy.special.i0! I also compared two more modern and simpler approximations that try to fit the modified Bessel function of the first kind and was disappointed in how poorly they approximated the function.

mruberry replied:
So then I read this blog post: https://www.advanpix.com/2015/11/11/rational-approximations-for-the-modified-bessel-function-of-the-first-kind-i0-computations-double-precision/#id2020094592. 

This was actually implemented in a modified form in Boost (see code links in the comments). I implemented their approach in Python (the correct argument is `x * x / 4` for the small argument approximation, despite the blog's Latex showing a ceil) and was very impressed with its accuracy. 

I actually really like this method because its derivation is clear and it's adaptable to different float types. How interested are you in the numerics? I think it'd be really interesting to investigate this approach further.

mruberry replied:
The copyright notice for Cephes we should use is:

> Copyright (c) 2018, Steven Moshier
> All rights reserved.
> 
> Redistribution and use in source and binary forms, with or without
> modification, are permitted provided that the following conditions are met:
>     * Redistributions of source code must retain the above copyright
>       notice, this list of conditions and the following disclaimer.
>     * Redistributions in binary form must reproduce the above copyright
>       notice, this list of conditions and the following disclaimer in the
>       documentation and/or other materials provided with the distribution.
>     * Neither the name of the <organization> nor the
>       names of its contributors may be used to endorse or promote products
>       derived from this software without specific prior written permission.
> 
> THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
> ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
> WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
> DISCLAIMED. IN NO EVENT SHALL Steven Moshier BE LIABLE FOR ANY
> DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
> (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
> LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
> ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
> (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
> SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Let's put this license in a note:

```
// Note [3-Clause BSD License for the Cephes Math Library]
// Code derived from implementations in the Cephes Math Library should mention its derivation and reference 
// this note (ex. 'This function is derived from the implementation of X in the Cephes Math Library. See note
// [3-Clause BSD License for the Cephes Math Library]. The license is:
// <license goes here>
```

Then each function can just explain its derivation and reference the note.

muthuArivoli replied:
Where would I place this note (what directory/file)? Also, should I change the comments on the zeta and digamma functions to follow this reference style for cephes?

mruberry replied:
You can replace this first comment in ATen/native/Math.h with it.

Thank you for asking. Yes, it'd be great if you would update the other Cephes references. Looks like there are existing references in ATen/native/cuda/Math.cuh, ATen/native/Math.h, and ATen/native/Distributions.h.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -384,3 +384,120 @@ calc_gcd(T a, T b) {
+}
+
+/*
+ * The following function has been modified from code that comes with the following copyright notice.
```
See note above. It's all Cephes-derived.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -384,3 +384,120 @@ calc_gcd(T a, T b) {
+ * Direct inquiries to 30 Frost Street, Cambridge, MA 02140
+ */
+template <typename T>
+static inline typename std::enable_if<std::is_floating_point<T>::value, T>::type
```
Comment explaining this function

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -384,3 +384,120 @@ calc_gcd(T a, T b) {
+ */
+template <typename T>
+static inline typename std::enable_if<std::is_floating_point<T>::value, T>::type
+calc_i0(T _x) {
```
comment explaining this function

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -984,6 +984,34 @@ Tensor hann_window(
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ kaiser_window ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor kaiser_window(int64_t window_length, double beta, const TensorOptions& options) {
+  return native::kaiser_window(window_length, beta, /*periodic=*/true, options);
```
Why include the periodic arg? Following https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.windows.kaiser.html?highlight=kaiser#scipy.signal.windows.kaiser, wouldn't it be more natural for the arg to be "sym" with a default of True? It's also fine to not expose this argument for now if it's unsupported.

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -984,6 +984,34 @@ Tensor hann_window(
+  auto window = native::arange(window_length, options);
+  auto alpha = (window_length - 1) / 2.0;
+  auto denom = native::full(window_length, beta, options);
+  window.sub_(alpha).div_(alpha).pow_(2.0).mul_(-1.0).add_(1.0).sqrt_().mul_(beta).i0_().div_(denom.i0_());
```
This is going to be a lot of kernel launches and redundant memory reads. What would you think about fusing these operations into a signal kernel for performance? If that seems like a lot, what about separating the i0 PR, which is incredibly interesting numerically, and the kaiser window PR, which requires fusion to be made performant?

mruberry commented on test/test_torch.py:
```diff
@@ -17257,6 +17264,38 @@ def test_nextafter(self, device, dtype):
+    @dtypesIfCPU(torch.bfloat16, torch.float32, torch.float64)
+    @dtypes(torch.float32, torch.float64)
+    @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    def test_i0(self, device, dtype):
```
This is a solid test but I'd like to hear your thinking about i0's numerics before discussing testing further.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -2990,6 +2990,29 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.i0,
+           r"""
+i0(input, out=None) -> Tensor
```
input, *, out

mruberry commented on torch/_torch_docs.py:
```diff
@@ -8390,6 +8413,49 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.kaiser_window,
+           """
+kaiser_window(window_length, beta, periodic=True, dtype=None, \
```
Mentioned this earlier, but the third parameter would be `sym=True` to be consistent with NumPy. The parameters from TensorOptions are also kwarg-only, right? So it'd be `sym=True, *, ...`

mruberry commented:
Hey @muthuArivoli! This is a really exciting PR with a super interesting numerical challenge in i0 and performantly implementing the Kaiser windowing function. If you're up for it, I'd like to propose separating the PRs and focusing on i0 first, and reviewing the approach in the article I linked and really focusing on the numerics of the function. In particular, in addition to comparing against SciPy's i0 I wonder if we should have a separate "golden values" test where we handcode a list of 20ish or more i0 values to compare with. I'm also curious to hear your thoughts about computing i0 in different floating point precisions.

Let me know if you're interested in investigating the numerics of i0 more and think that approach makes sense. The Cephes-based implementation is certainly high quality, although it's interesting that SciPy's i0 appears to be promoting half->float. From your tests it doesn't appear that this is actually a numerics issue, though?

muthuArivoli commented:
I'd be happy to look at different numerical approaches. I'll go ahead and remove the kaiser window stuff from this PR. From my understanding, during the calculation, Scipy promotes everything to double(float64) and then casts the final answer back down. I'm not sure if it is numerics, but I was getting incorrect answers in my initial implementation that used the input type for the types of the constants. I probably needed to remove the constants that were too small for the input dtype.

mruberry commented:
> I'd be happy to look at different numerical approaches. I'll go ahead and remove the kaiser window stuff from this PR. From my understanding, during the calculation, Scipy promotes everything to double(float64) and then casts the final answer back down. I'm not sure if it is numerics, but I was getting incorrect answers in my initial implementation that used the input type for the types of the constants. I probably needed to remove the constants that were too small for the input dtype.

Or, following the discussion in the blog, maybe there are better per-dtype constants. Open questions:

- Following the blog post, which approach to computing i0 is most accurate, which is fastest?
- Which generalizes to other dtypes best? Is computing the function in lower precision floats, like float32, float16, or bfloat16 realistic?

mruberry commented:
Code I've been using to interrogate the different methods:

"""
import numpy as np
import scipy
import scipy.special as special
import math
import decimal
np.set_printoptions(50)

v = 4.990

# Only the small value approach from the blog is implemented below
assert v < 7.75

def approx(a, i=100):
  a = decimal.Decimal(a)
  sum = decimal.Decimal(0)
  fac = 1
  quarter = decimal.Decimal(1/4)
  for k in range(i):
    result = ((quarter * (a**2))**k) / (fac * fac)
    sum += result

    fac *= k + 1

  return sum

def p(x):
  y = 1.0000000000000000000000801e+00 + 2.4999999999999999999629693e-01 * x + 2.7777777777777777805664954e-02 * (x**2) + \
  1.7361111111111110294015271e-03 * (x**3) + 6.9444444444444568581891535e-05 * (x**4) + 1.9290123456788994104574754e-06 * (x**5) + \
  3.9367598891475388547279760e-08 * (x**6) + 6.1511873265092916275099070e-10 * (x**7) + 7.5940584360755226536109511e-12 * (x**8) + \
  7.5940582595094190098755663e-14 * (x**9) + 6.2760839879536225394314453e-16 * (x**10) + 4.3583591008893599099577755e-18 * (x**11) + \
  2.5791926805873898803749321e-20 * (x**12) + 1.3141332422663039834197910e-22 * (x**13) + 5.9203280572170548134753422e-25 * (x**14) + \
  2.0732014503197852176921968e-27 * (x**15) + 1.1497640034400735733456400e-29 * (x**16)

  return y

def blog(x):
  y = x * x /4
  py = p(y)
  return 1 + y * py

approx_result = approx(v, 400)   
scipy_result = special.i0(np.array((v,), dtype=np.float64))
blog_result = blog(v)

approx_vs_scipy = abs(approx_result - decimal.Decimal(scipy_result.item()))
approx_vs_blog = abs(approx_result - decimal.Decimal(blog_result))
print("Approx vs SciPy: ", approx_vs_scipy)
print("Approx vs Blog: ", approx_vs_blog)
print("Smaller error is: ", "Scipy" if approx_vs_scipy < approx_vs_blog else "Blog")
"""

muthuArivoli commented:
Here are my benchmarking results. To evaluate accuracy, I compared the results with Mathematica, not sure if that's the best way to go, but I was hoping they would have the most accurate results, since they do symbolic math. Overall, I found that the Boost implementation is 3-10x faster than Cephes, but Cephes is closer to Mathematica by an order of magnitude. Accuracy was tested for input over a uniform distribution in range [0, 30]. I have also included results from numpy for accuracy only. Here are the results:
```
Cephes Implementation
Speed
dtype=torch.float32, size=1000, multiplier=7.75 time_per_iter=0.00011358737945556641
dtype=torch.float32, size=1000, multiplier=15 time_per_iter=0.00011703014373779297
dtype=torch.float32, size=1000, multiplier=30 time_per_iter=0.00011022090911865235
dtype=torch.float64, size=1000, multiplier=7.75 time_per_iter=0.00011109352111816406
dtype=torch.float64, size=1000, multiplier=15 time_per_iter=0.0001123952865600586
dtype=torch.float64, size=1000, multiplier=30 time_per_iter=0.00010863304138183593
dtype=torch.float32, size=10000, multiplier=7.75 time_per_iter=0.0015386724472045898
dtype=torch.float32, size=10000, multiplier=15 time_per_iter=0.0008890628814697266
dtype=torch.float32, size=10000, multiplier=30 time_per_iter=0.0006285333633422852
dtype=torch.float64, size=10000, multiplier=7.75 time_per_iter=0.000753021240234375
dtype=torch.float64, size=10000, multiplier=15 time_per_iter=0.0006937932968139648
dtype=torch.float64, size=10000, multiplier=30 time_per_iter=0.0006339120864868164
dtype=torch.float32, size=100000, multiplier=7.75 time_per_iter=0.006505584716796875
dtype=torch.float32, size=100000, multiplier=15 time_per_iter=0.006360173225402832
dtype=torch.float32, size=100000, multiplier=30 time_per_iter=0.0062765836715698246
dtype=torch.float64, size=100000, multiplier=7.75 time_per_iter=0.006743760108947754
dtype=torch.float64, size=100000, multiplier=15 time_per_iter=0.006422591209411621
dtype=torch.float64, size=100000, multiplier=30 time_per_iter=0.00616114616394043
dtype=torch.float32, size=1000000, multiplier=7.75 time_per_iter=0.06382527828216553
dtype=torch.float32, size=1000000, multiplier=15 time_per_iter=0.06363551616668701
dtype=torch.float32, size=1000000, multiplier=30 time_per_iter=0.06286612510681153
dtype=torch.float64, size=1000000, multiplier=7.75 time_per_iter=0.06520251274108886
dtype=torch.float64, size=1000000, multiplier=15 time_per_iter=0.06453951358795167
dtype=torch.float64, size=1000000, multiplier=30 time_per_iter=0.06095137596130371
dtype=torch.float32, size=10000000, multiplier=7.75 time_per_iter=0.6557976293563843
dtype=torch.float32, size=10000000, multiplier=15 time_per_iter=0.6597203779220581
dtype=torch.float32, size=10000000, multiplier=30 time_per_iter=0.6464773988723755
dtype=torch.float64, size=10000000, multiplier=7.75 time_per_iter=0.6917289638519287
dtype=torch.float64, size=10000000, multiplier=15 time_per_iter=0.6825681209564209
dtype=torch.float64, size=10000000, multiplier=30 time_per_iter=0.6503863286972046

Accuracy
mean: -40.74425506591797, rms: 54998.6484375, rel dif: 3.149424054882388e-10
mean: -4.4761938905600586e-07, rms: 2.2956596935970257e-05, rel dif: 8.130645209920435e-17



Boost Implementation
Speed
dtype=torch.float32, size=1000, multiplier=7.75 time_per_iter=1.0471343994140625e-05
dtype=torch.float32, size=1000, multiplier=15 time_per_iter=1.0046958923339844e-05
dtype=torch.float32, size=1000, multiplier=30 time_per_iter=1.1458396911621094e-05
dtype=torch.float64, size=1000, multiplier=7.75 time_per_iter=1.35040283203125e-05
dtype=torch.float64, size=1000, multiplier=15 time_per_iter=2.5482177734375e-05
dtype=torch.float64, size=1000, multiplier=30 time_per_iter=3.0231475830078124e-05
dtype=torch.float32, size=10000, multiplier=7.75 time_per_iter=4.912853240966797e-05
dtype=torch.float32, size=10000, multiplier=15 time_per_iter=9.809017181396485e-05
dtype=torch.float32, size=10000, multiplier=30 time_per_iter=8.341789245605469e-05
dtype=torch.float64, size=10000, multiplier=7.75 time_per_iter=9.78708267211914e-05
dtype=torch.float64, size=10000, multiplier=15 time_per_iter=0.0002141237258911133
dtype=torch.float64, size=10000, multiplier=30 time_per_iter=0.00025643348693847655
dtype=torch.float32, size=100000, multiplier=7.75 time_per_iter=0.00044190406799316407
dtype=torch.float32, size=100000, multiplier=15 time_per_iter=0.0010473203659057617
dtype=torch.float32, size=100000, multiplier=30 time_per_iter=0.0008917665481567383
dtype=torch.float64, size=100000, multiplier=7.75 time_per_iter=0.000940098762512207
dtype=torch.float64, size=100000, multiplier=15 time_per_iter=0.002071399688720703
dtype=torch.float64, size=100000, multiplier=30 time_per_iter=0.002510538101196289
dtype=torch.float32, size=1000000, multiplier=7.75 time_per_iter=0.00433619499206543
dtype=torch.float32, size=1000000, multiplier=15 time_per_iter=0.010205540657043457
dtype=torch.float32, size=1000000, multiplier=30 time_per_iter=0.008591766357421876
dtype=torch.float64, size=1000000, multiplier=7.75 time_per_iter=0.008778400421142578
dtype=torch.float64, size=1000000, multiplier=15 time_per_iter=0.021117911338806153
dtype=torch.float64, size=1000000, multiplier=30 time_per_iter=0.02476226806640625
dtype=torch.float32, size=10000000, multiplier=7.75 time_per_iter=0.05833768844604492
dtype=torch.float32, size=10000000, multiplier=15 time_per_iter=0.11765189647674561
dtype=torch.float32, size=10000000, multiplier=30 time_per_iter=0.10121417999267578
dtype=torch.float64, size=10000000, multiplier=7.75 time_per_iter=0.12034669399261474
dtype=torch.float64, size=10000000, multiplier=15 time_per_iter=0.24495739459991456
dtype=torch.float64, size=10000000, multiplier=30 time_per_iter=0.28176236629486084

Accuracy
mean: -916.7548217773438, rms: 55375.46484375, rel dif: -2.5779944845538694e-09
mean: 3.0042552458229865e-06, rms: 2.7350977682692713e-05, rel dif: 1.479971044306233e-16

Numpy Implementation
Accuracy
mean: -1019.195556640625, rms: 55555.9140625, rel dif: -9.79536629586164e-09
mean: -4.4834493140177176e-07, rms: 2.2958003553987907e-05, rel dif: 8.110220616586299e-17
```
Python code
```
import torch
import time
import numpy as np
import scipy.special

torch.set_num_threads(1)
iters = 50

print("Speed")
for i in [1000, 10000, 100000, 1000000, 10000000]:
    for dtype in [torch.float32, torch.float64]:
        for multiplier in [7.75, 15, 30]:
            a = torch.rand(i, dtype=dtype) * multiplier

            start_time = time.time()
            for iter in range(iters):
                x = a.i0()
            total_time = time.time() - start_time
            time_per_iter = total_time / iters
            print("dtype={}, size={}, multiplier={} time_per_iter={}".format(dtype, i, multiplier, time_per_iter))

print("Accuracy")
for type in [('input32.bin', 'float32'), ('input64.bin', 'float64')]:
    with open(type[0], 'rb') as f:
        inp = np.fromfile(f, dtype=type[1])
        torch_ans = torch.i0(torch.from_numpy(inp))
        with open("output.bin", "rb") as outf:
            mathematica_ans = np.fromfile(outf, dtype='float64').astype(type[1])
            difference = torch_ans - torch.from_numpy(mathematica_ans)
            mean = torch.mean(difference)
            rms = np.sqrt(torch.sum(difference**2)/np.size(inp))
            rel_dif = torch.mean(difference/torch.from_numpy(mathematica_ans))
            print('mean: {}, rms: {}, rel dif: {}'.format(mean, rms, rel_dif))
```
Mathematica notebook
```
x = RandomReal[30, 1000000]

f = OpenWrite["input64.bin", BinaryFormat -> True];
BinaryWrite["input64.bin", x, "Real64"]
Close[f]
f2 = OpenWrite["input32.bin", BinaryFormat -> True];
BinaryWrite["input32.bin", x, "Real32"]
Close[f2]

y = BesselI[0, x]

g = OpenWrite["output.bin", BinaryFormat -> True];
BinaryWrite["output.bin", y, "Real64"]
Close[g]
```

muthuArivoli commented:
If I take the absolute value of the difference, `difference = torch.abs(torch_ans - torch.from_numpy(mathematica_ans))`, I get the following
```
Cephes
mean: 12376.255859375, rms: 54998.6484375, rel dif: 2.960472329505137e-07
mean: 4.660421135407815e-06, rms: 2.2958003553987907e-05, rel dif: 2.5862337122366687e-16

Boost
mean: 12425.150390625, rms: 55375.46484375, rel dif: 3.004124664585106e-07
mean: 5.281933232628967e-06, rms: 2.7350977682692713e-05, rel dif: 2.6849518849477342e-16

Numpy
mean: 12454.1533203125, rms: 55555.9140625, rel dif: 3.101684171724628e-07
mean: 4.660421135407815e-06, rms: 2.2958003553987907e-05, rel dif: 2.5862337122366687e-16
```
This now makes me think that the Boost implementation has nearly the same accuracy as Cephes.

mruberry commented:
My tests on 0-7.75 also suggest that the Cephes-based approach has a small accuracy advantage when computing in double. The average error for both approaches in this range was less than 1e-14.

I also followed-up on licensing and we can derive from either approach without issue.

Going with the more accurate approach for a function like this sounds good, and as a bonus you've already done a lot of the work for a Cephes-based approach. 

My next numerics question, then, is what kind of degradation do we see in the numerics of the Cephes-based approach if we compute in lower precision types like float16, bfloat16, and float32?

mruberry commented:
Also, we should include some larger values in the analysis (the following are picked arbitrarily):

73 = 2.363168219162483471104054117E+30
102.35 = 1.112857063386947721952064289E+43
147.77 = 4.922323936053304677568747730E+62
203.99 = 1.091701157563273761572824616E+87
313.01 = 1.958035010023479793955068216E+134

These values are computed using an approximating series, and the absolute error between them and SciPy is huge. The relative error, however, is small. We can also compare the float to the double values and look at the relative error for switching to float from double by just comparing those results for consistency.

muthuArivoli commented:
Computing in float32 with Cephes does give me large absolute errors (the first line for each library in the data above for accuracy is using float32), but the max relative error I've seen is in the 1e-7 range (when comparing with the result I would have gotten had I used doubles). For bfloat16 and float16, I think those would need to be cast to a regular float, since I remember I was getting large errors that would simply fail the tests with standard rtol and atol values, if I didn't cast them up, even for values the [0,8] range.  

> 73 = 2.363168219162483471104054117E+30
> 102.35 = 1.112857063386947721952064289E+43
> 147.77 = 4.922323936053304677568747730E+62
> 203.99 = 1.091701157563273761572824616E+87
> 313.01 = 1.958035010023479793955068216E+134

For these values, the last 4 give me inf when using float32, so I'm not sure how to use them to test between float and double. The first value gave me an error of 0.

mruberry commented:
> For these values, the last 4 give me inf when using float32, so I'm not sure how to use them to test between float and double. The first value gave me an error of 0.

You're right, my mistake: those values are (mostly) too big for float32. The first one is representable, though. Can you elaborate on the issue you're seeing?

muthuArivoli commented:
> You're right, my mistake: those values are (mostly) too big for float32. The first one is representable, though. Can you elaborate on the issue you're seeing?

Sorry, I meant no issue. The float value was exactly equal to the double value which was equal to the value that you gave as the answer.
 
> Can you detail what a "large error" is? Like, what is the typical relative error for bfloat16 and float16 on values in the 0-10 range?

For bfloat16, I'm seeing relative errors in the 1e-3 range. I can't test float16, since I don't have a gpu, but I'll push code soon to test that out on the ci.
This value seems fine, so I'm guessing I was incorrect upcasting something in my test that caused a lower tolerance to be used.

muthuArivoli committed:
remove kaiser window

mruberry commented:
> > You're right, my mistake: those values are (mostly) too big for float32. The first one is representable, though. Can you elaborate on the issue you're seeing?
> 
> Sorry, I meant no issue. The float value was exactly equal to the double value which was equal to the value that you gave as the answer.
> 
> > Can you detail what a "large error" is? Like, what is the typical relative error for bfloat16 and float16 on values in the 0-10 range?
> 
> For bfloat16, I'm seeing relative errors in the 1e-3 range. I can't test float16, since I don't have a gpu, but I'll push code soon to test that out on the ci.
> This value seems fine, so I'm guessing I was incorrectly upcasting something in my test that caused a lower tolerance to be used.

Errors like 1e-3 on the 0-10 interval do seem OK, and we can use the `@precisionOverrides` decorator to specify per-dtype tolerances to make the test pass if we need to. I am curious what the relative error of bfloat16 and float16 on higher values are, though. Maybe in the test we can use SciPy as a reference and have different values to compare with for each dtype (so we don't hit inf issues)?

muthuArivoli committed:
template for other dtypes

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-kaiser

muthuArivoli committed:
fix comments and docs

muthuArivoli committed:
add ranged tests

muthuArivoli commented:
I partitioned the test set into categories depending on the max value that could be used. float16 seems to do fine. The bfloat16 test is failing for just 2 out of a 1000 elements on Windows and Mac, but not on Linux, so I'm not sure what to make of that.

mruberry commented:
> I partitioned the test set into categories depending on the max value that could be used. float16 seems to do fine. The bfloat16 test is failing for just 2 out of a 1000 elements on Windows and Mac, but not on Linux, so I'm not sure what to make of that.

Interesting. Would partitioning into separate tests with different atol overrides using @precisionOverrides make sense to handle the bfloat16 failures?

muthuArivoli committed:
use precision override

muthuArivoli commented:
> Interesting. Would partitioning into separate tests with different atol overrides using @precisionOverride make sense to handle the bfloat16 failures?

Thanks, I set the precisionOverride to 3e-2, and that worked. Are there any specific values that I should hardcode and test against? If not, I think this should be ready for another review.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -384,3 +384,137 @@ calc_gcd(T a, T b) {
+ * Cephes Math Library Release 2.8:  June, 2000
+ * Copyright 1984, 1987, 2000 by Stephen L. Moshier
+ *
+ * Returns modified Bessel function of order zero of the argument.
```
"zeroth order modified Bessel function of the first kind"

mruberry replied:
(Should probably also mention that it takes the abs of negative inputs).

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -196,5 +196,144 @@ static inline C10_HOST_DEVICE scalar_t calc_gcd(scalar_t a_in, scalar_t b_in) {
 }
 
+/*
+ * The following function was converted to CUDA form from code that comes
```
This section, for example, could reference the Cephes note

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -196,5 +196,144 @@ static inline C10_HOST_DEVICE scalar_t calc_gcd(scalar_t a_in, scalar_t b_in) {
+ * Copyright 1984, 1987, 2000 by Stephen L. Moshier
+ *
+ * Returns modified Bessel function of order zero of the argument.
+ * The range is partitioned into the two intervals [0,8] and (8, infinity). Chebyshev polynomial expansions are employed
```
The CUDA functions can also point to the CPU function documentations if you want to avoid redundancy and prevent divergence

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -196,5 +196,144 @@ static inline C10_HOST_DEVICE scalar_t calc_gcd(scalar_t a_in, scalar_t b_in) {
+template <typename scalar_t>
+static inline C10_HOST_DEVICE scalar_t calc_i0(scalar_t _x) {
+  using accscalar_t = at::acc_type<scalar_t, true>;
+  accscalar_t x = ::abs(static_cast<accscalar_t>(_x));
```
This will perform the computation in float32 with bfloat16 and float16 inputs, right?

That may be the correct thing to do, but before jumping to that conclusion we should look at the numerical accuracy (or inaccuracy) of computing in bfloat16/float16 first.

mruberry replied:
Update: Same comment as above explaining the upcasting would go here, too

mruberry commented on test/test_torch.py:
```diff
@@ -16626,6 +16630,65 @@ def test_nextafter(self, device, dtype):
+        if dtype is torch.bfloat16:
+            a = a.to(torch.float32)
+        expected = scipy.special.i0(a.cpu().numpy())
+        if dtype is torch.bfloat16 or dtype is torch.float16:
```
why the float16 check here?

mruberry replied:
Answering my own question: because of the upcasting that SciPy does. A comment would be helpful.

mruberry commented on test/test_torch.py:
```diff
@@ -16626,6 +16630,65 @@ def test_nextafter(self, device, dtype):
+            expected = torch.from_numpy(expected).to(dtype)
+        self.assertEqual(actual, expected)
+
+        a = torch.rand(100, device=device).to(dtype) * -13.25
```
Better merge this block with the previous one:
```
t_pos = torch.rand(100, device=device).to(dtype) * 13.25
t_neg = torch.rand(100, device=device).to(dtype) * -13.25
for t in (t_pos, t_neg):
  ...
```

mruberry commented on test/test_torch.py:
```diff
@@ -16626,6 +16630,65 @@ def test_nextafter(self, device, dtype):
         expected = np.nextafter(a.cpu().numpy(), b.cpu().numpy())
         self.assertEqual(actual, expected, atol=0, rtol=0)
 
+    @dtypesIfCUDA(torch.float16, torch.float32, torch.float64)
```
Add a comment explaining this test (and the domain it's testing)

mruberry commented on test/test_torch.py:
```diff
@@ -16626,6 +16630,65 @@ def test_nextafter(self, device, dtype):
+            expected = torch.from_numpy(expected).to(dtype)
+        self.assertEqual(actual, expected)
+
+    @dtypesIfCPU(torch.bfloat16, torch.float32, torch.float64)
```
Comment explaining test + domain

mruberry commented on test/test_torch.py:
```diff
@@ -16626,6 +16630,65 @@ def test_nextafter(self, device, dtype):
+    @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    def test_i0_range2(self, device, dtype):
+        a = torch.rand(1000, device=device).to(dtype) * 88.5
+        actual = torch.i0(a)
```
Since the same block appears again here, you may want to refactor it into a helper function:

`def _i0_helper(t): ...`

Then you can just compute the tensors and call the helper function in these tests.

mruberry commented on test/test_torch.py:
```diff
@@ -16626,6 +16630,65 @@ def test_nextafter(self, device, dtype):
+            expected = torch.from_numpy(expected).to(dtype)
+        self.assertEqual(actual, expected)
+
+    @dtypes(torch.float64)
```
Comment + domain explanation + use helper function

mruberry commented on test/test_torch.py:
```diff
@@ -16626,6 +16630,65 @@ def test_nextafter(self, device, dtype):
+        actual = torch.i0(a)
+        expected = scipy.special.i0(a.cpu().numpy())
+        self.assertEqual(actual, expected)
+
```
+ A "corner case" test for handling -inf, inf, and nan inputs as well as empty tensor inputs.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3098,6 +3098,29 @@ def merge_dicts(*dicts):
+           r"""
+i0(input, *, out=None) -> Tensor
+
+Computes the modified Bessel function of the first kind, order 0, for each element of :attr:`input`.
```
For consistency, this might become "Computes the zeroth order modified Bessel function of the first kind for each element of :attr:\`input\`."

mruberry reviewed (changes requested):
Overall this looks really excellent and I appreciate the time we took to investigate the numerics of this function.

Currently I think the CUDA kernel is computing in float32 when given half inputs. Would you check, and, if it is, check the numerical accuracy when computing in float16? Maybe float16 should compute in float32, but maybe, like bfloat16, it's OK not to.

I also made a few comments about documentation, licensing, and simplifying and extending the tests.

muthuArivoli committed:
licensing and documentation updates

muthuArivoli committed:
compute in given type

muthuArivoli committed:
update tests

muthuArivoli commented:
It seems like there actually is an accuracy problem when using bfloat16 or float16. When I increased the size of the tensor to a 1000, both the bfloat16 and float16 tests fail, with a max error of 4. I'm not entirely sure why this is occurring when I increase the size to 1000 - I would expect a random sample of 100 values to fail, especially since the input is only in range [0, 13.25). I'm also not sure why this is only occurring on Windows/Mac, but works fine on Linux, I would expect it to also fail on Linux as well. Should I change the code back to using float32 for bfloat16 and float16?

muthuArivoli committed:
Revert "compute in given type"

This reverts commit 008248d01f385ead8c8c1841913500aa7f5af056.

muthuArivoli committed:
compute in float for bfloat

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-kaiser

mruberry commented:
> It seems like there actually is an accuracy problem when using bfloat16 or float16. When I increased the size of the tensor to a 1000, both the bfloat16 and float16 tests fail, with a max error of 4. I'm not entirely sure why this is occurring when I increase the size to 1000 - I would expect a random sample of 100 values to fail, especially since the input is only in range [0, 13.25). I'm also not sure why this is only occurring on Windows/Mac, but works fine on Linux, I would expect it to also fail on Linux as well. Should I change the code back to using float32 for bfloat16 and float16?

Interesting! There are a few options:

1) you can set the precision conditionally for Mac and Windows (ex. `@precisionOverride({torch.bfloat16: bfloat16_prec})` where `bfloat16_prec` is computed used flags like `IS_WINDOWS` and `IS_MACOS` (I think that's the name of those flags, I may be misremembering) and add a warning in the documentation to use at least float32 on mac and windows
2) we could always compute in at least float32 across all platforms - but it sounds like the error for float16 and bfloat16 are low on linux

If we're confident that float16 and bfloat16 are accurate on Linux then I'd like to suggest trying to preserve the dtype.

mruberry commented on aten/src/ATen/core/aten_interned_strings.h:
```diff
@@ -373,6 +373,7 @@ _(aten, histc) \
 _(aten, hspmm) \
 _(aten, hstack) \
 _(aten, hypot) \
+_(aten, i0) \
```
and `_(aten, i0_)`

mruberry commented on aten/src/ATen/native/Distributions.h:
```diff
@@ -259,11 +259,8 @@ C10_DEVICE scalar_t sample_binomial(scalar_t count, scalar_t prob, BaseSampler<a
- * Cephes Math Library Release 2.8:  June, 2000
- * Copyright 1984, 1987, 1992, 2000 by Stephen L. Moshier
+ * This function is derived from the implementation of the digamma function in the Cephes Math Library.
+ * See note [3-Clause BSD License for the Cephes Math Library] in "ATen/native/Math.h".
```
No need for quotes around the path.

mruberry commented on test/test_torch.py:
```diff
@@ -16872,6 +16876,55 @@ def test_nextafter(self, device, dtype):
         expected = np.nextafter(a.cpu().numpy(), b.cpu().numpy())
         self.assertEqual(actual, expected, atol=0, rtol=0)
 
+    def _i0_helper(self, t, dtype):
```
Nit: this signature could be simplified by acquiring the dtype from t itself (`t.dtype`)

mruberry commented on test/test_torch.py:
```diff
@@ -16872,6 +16876,55 @@ def test_nextafter(self, device, dtype):
+
+    @dtypes(torch.float64)
+    @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    def test_i0_range3(self, device, dtype):
```
Really nice test granularity and use of helpers in these tests.

mruberry commented:
Just did a quick scan. PR looks excellent. I think we just want to resolve the last design decision of how to compute for bfloat16 and float16. Let me try to enumerate the options:

1) just compute in them unconditionally (but maybe document that the functions are inaccurate on Mac and Windows)
2) compute in float32 but cast the intermediate values back to float16 or bfloat16 (i.e. float16 -> float32 computation -> float16)
3) compute in float32 and return a result in float32 (like SciPy does)

Option (2) seems the safest to me. PyTorch likes to preserve dtypes where possible, and computing in float32 is easy to do (it would also let you support bfloat16 and half on both CPU and CUDA, I think, and without a custom bfloat16 vectorized implementation, even though that's very cool). Option (1) is better if it's actually faster and preserves accuracy across platforms, however. 

What are your thoughts?

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-kaiser

muthuArivoli committed:
review fixes

muthuArivoli commented:
> Just did a quick scan. PR looks excellent. I think we just want to resolve the last design decision of how to compute for bfloat16 and float16. Let me try to enumerate the options:
> 
> 1. just compute in them unconditionally (but maybe document that the functions are inaccurate on Mac and Windows)
> 2. compute in float32 but cast the intermediate values back to float16 or bfloat16 (i.e. float16 -> float32 computation -> float16)
> 3. compute in float32 and return a result in float32 (like SciPy does)
> 
> Option (2) seems the safest to me. PyTorch likes to preserve dtypes where possible, and computing in float32 is easy to do (it would also let you support bfloat16 and half on both CPU and CUDA, I think, and without a custom bfloat16 vectorized implementation, even though that's very cool). Option (1) is better if it's actually faster and preserves accuracy across platforms, however.
> 
> What are your thoughts?

I agree, I think that option 2 is safer. Especially since other functions are going to be built on top of this function, an error here could propagate and could have unintended effects on those functions. I also think that the error of 4.0 is unacceptable when computing in the lower dtypes, and since it seems to only come out on certain random values spread through the domain, that could be very hard for people to predict against. It could also possibly cost people more time to cast their tensor up and then back down, if they are on Mac or Windows, so while people on Linux would see a performance gain, those on Mac/Windows would see a loss. 

I think this could use another review now, I believe I have addressed all of the existing review comments.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -384,3 +405,137 @@ calc_gcd(T a, T b) {
+  return static_cast<T>(std::exp(x) * chbevl(static_cast<T>(32.0 / x - 2.0), B, 25) / std::sqrt(x));
+}
+
+inline c10::BFloat16 calc_i0(c10::BFloat16 a) { return calc_i0(static_cast<float>(a)); }
```
Let's add a comment explaining the upcasting behavior

mruberry commented on aten/src/ATen/native/cuda/UnaryOpsKernel.cu:
```diff
@@ -47,6 +47,14 @@ void expm1_kernel_cuda(TensorIterator& iter) {
 }
 
+void i0_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "i0_cuda", [&]() {
```
Could we add bfloat16 here and see it just work since bfloat16 uses float32 as its accumulate type?

https://github.com/pytorch/pytorch/blob/324c18fcad579b1afa63ae45528bf598ba8ec4ca/aten/src/ATen/AccumulateType.h#L25

mruberry commented on test/test_torch.py:
```diff
@@ -16872,6 +16876,56 @@ def test_nextafter(self, device, dtype):
         expected = np.nextafter(a.cpu().numpy(), b.cpu().numpy())
         self.assertEqual(actual, expected, atol=0, rtol=0)
 
+    def _i0_helper(self, t):
```
Brief comment here explaining the i0 tests and why they cut up the range.

mruberry reviewed (approved):
This is awesome! I left a few requests for tiny comments so that future readers will have an easier time understanding the code and tests. Just ping me when this is ready.

muthuArivoli committed:
updates

muthuArivoli committed:
fix comments

muthuArivoli commented:
@mruberry, I've added the comments and added bfloat16 for GPU. The ROCm test failed for some unrelated reason, so I'm not sure if bfloat16 actually works or not, but other than that I think this should be good to go.
